src/runtime/StateSerialization/RuntimeData.cs

```diff
@@ -250,11 +280,82 @@ private static void RestoreRuntimeDataObjects(SharedObjectsState storage)
             }
         }
 
+        /// <summary>
+        /// Frees the pointer stored in a Python capsule and the capsule stored 
+        /// on the sys module object with the given key if it exists. 
+        /// </summary>
+        /// <remarks>
+        /// The memory on the capsule must have been allocated via <code>StashDataInCapsule</code>
+        /// </remarks>
+        /// <param name="key">The name given to the capsule on the `sys` module object</param>
+        public static void FreeCapsuleData(string key)
+        {
+            BorrowedReference oldCapsule = PySys_GetObject(key);
+            if (!oldCapsule.IsNull)
+            {
+                IntPtr oldData = PyCapsule_GetPointer(oldCapsule, IntPtr.Zero);
+                Marshal.FreeHGlobal(oldData);
+                PyCapsule_SetPointer(oldCapsule, IntPtr.Zero);
+                PySys_SetObject(key, null);
+            }
+        }
+
+        /// <summary>
+        /// Stores the <paramref name="data"/>data parameter in a Python capsule and stores 
+        /// the capsule on the `sys` module object with the name <paramref name="key"/>. 
+        /// This method allocates global memory to hold the data of the <paramref name="data"/>
+        /// parameter.
+        /// </summary>
+        /// <remarks>
+        /// No checks on pre-existing names on the `sys` module object are made.
+        /// </remarks>
+        /// <param name="key">The name given to the capsule on the `sys` module object</param>
+        /// <param name="data">The data to be contained in the capsule</param>
+        public static void StashDataInCapsule(string key, byte[] data)
+        {
+            IntPtr mem = Marshal.AllocHGlobal(IntPtr.Size + data.Length);
+            // store the length of the buffer first
+            Marshal.WriteIntPtr(mem, (IntPtr)data.Length);
+            Marshal.Copy(data, 0, mem + IntPtr.Size, data.Length);
+
+            using NewReference capsule = PyCapsule_New(mem, IntPtr.Zero, IntPtr.Zero);
+            int res = PySys_SetObject(key, capsule.BorrowOrThrow());
+            PythonException.ThrowIfIsNotZero(res);
+        }
+
+        /// <summary>
+        /// Retreives the pointer to previously stored data on a Python capsule. 
+        /// Throws if the object corresponding to the <paramref name="key"/> parameter
+        /// on the `sys` module object is not a capsule.
+        /// </summary>
+        /// <param name="key">The name given to the capsule on the `sys` module object</param>
+        /// <returns>The pointer to the data, or IntPtr.Zero if name matches the key</returns>
+        public static IntPtr GetDataFromCapsule(string key)
```

Comment:
While this works, consider providing higher-level interface.

I think something like a `MemoryStream` or `Span` would be better.

Might need to keep a `PyObject` around to ensure capsule is not destroyed while being accessed.

Reply:
As-is there's no way for the consumer to know the pointer points to data that starts with the memory block length.

Reply:
> Might need to keep a PyObject around to ensure capsule is not destroyed while being accessed.

I guess this could happen if a user calls `FreeCapsuleData` before using the data. Deallocating the capsule using `FreeCapsuleData` would not deallocate the data as we do not implement capsule destructors. I feel that's more a usage/documentation problem.
Otherwise it would mean returning a copy of the data, or the capsule itself.

Reply:
A minimal wrapper for the capsule type might make sense here.

Reply:
I have been thinking about it and having a wrapper over the capsule type does not circumvent the "calling FreeCapsuleData before using the data" scenario. The capsule does not own its data unless there is a destructor function set on it; tp_dealloc calls it if it exists (calling tp_dealloc on the capsule would leak the memory if no one frees the pointer in this scenario). Short of emitting assembly code for the destructor, I am not sure how we can accomplish the ownership of the data by the capsule.

Reply:
It does not have to be a generic capsule type, just something for this specific use case.

Reply:
I opted for using `MemoryStream` for both Stash and Get functions, as to give symmetry in the API.